### PR TITLE
[YUNIKORN-2292] Flaky E2E Test: Orphan pods still exist after TearDownNamespace()

### PR DESCRIPTION
**Pull request by chenyulin0719**
### What is this PR for?
- Check and delete Jobs in TearDownNamespace()

2024/01/02 Update:
- Also clenaup Deployment,DaemonSet,ReplicaSet,StatefulSet,PVC in TearDownNamespace()

### What type of PR is it?
* [x] - Bug Fix
* [ ] - Improvement
* [ ] - Feature
* [ ] - Documentation
* [ ] - Hot Fix
* [ ] - Refactoring

### Todos
NA

### What is the Jira issue?
https://issues.apache.org/jira/browse/YUNIKORN-2292

### How should this be tested?
All existing test should pass.
> make test
make e2e_test

### Screenshots (if appropriate)
NA

### Questions:
NA

### CODE REVIEW

**Comment by chenyulin0719** [comment]
Have checked the error message is equal to [YUNIKORN-2291 Flaky e2e test](https://issues.apache.org/jira/browse/YUNIKORN-2291)

One potential issue involves checking the application's failing state. I will create another Jira ticket once this issue is confirmed.
https://github.com/apache/yunikorn-k8shim/blob/master/test/e2e/gang_scheduling/gang_scheduling_test.go#L288

**Comment by chia7712** [comment]
@chenyulin0719 Is the failed e2e related to this PR?

File: test/e2e/framework/helpers/k8s/k8s_utils.go [hunk 1]
```diff
@@ -391,7 +391,13 @@ func (k *KubeCtl) DeleteNamespace(namespace string) error {
 }
 
 func (k *KubeCtl) TearDownNamespace(namespace string) error {
```
Comment (chia7712) [inline test/e2e/framework/helpers/k8s/k8s_utils.go:393]: Should it cleanup both PV and PVC ?

Comment (chenyulin0719) [reply test/e2e/framework/helpers/k8s/k8s_utils.go:393]: Hi @chia7712,
I don't think we should cleanup both PV and PVC in TearDownNamespace.

PV/PVC were only used in "persistent_volume" e2e test.  I think the cleanup responsibility is on the e2e test itself. (Besides, PV is cluster resource and is not bounded to namesapce.)

One of the reason we cleanup Jobs/Pods in TearDownNamespace() is those objects are widely used in serveral e2e tests and sadly we didn't handle all the cleanup.

(IMO, the best practice is each e2e test clean Jobs/Pods by themself. However, it quite difficult to review all the e2e test.)

**Comment by chenyulin0719** [comment]
> @chenyulin0719 Is the failed e2e related to this PR?

Hi @chia7712 ,
The failed e2e test is not related to this PR. I've created below Jira to cover it:

- [[YUNIKORN-2293] Flaky E2E Test: Failed asserts in LogTestClusterInfoWrapper() blocked the resources cleanup steps](https://issues.apache.org/jira/browse/YUNIKORN-2293)
- [[YUNIKORN-2294] Flaky E2E Test: "Verify_Hard_GS_Failed_State" polling short-lived "Failing" application status](https://issues.apache.org/jira/browse/YUNIKORN-2294)

File: test/e2e/framework/helpers/k8s/k8s_utils.go (hunk 1 above)
Comment (chia7712) [reply test/e2e/framework/helpers/k8s/k8s_utils.go:393]: My point was that all e2e should reuse the code to cleanup resources. PV, PVC, pods, and jobs should be cleanup when the e2e is closing even if the e2e does not use such resource. That could make isolation between e2e test and all tests get benefit from it.

Comment (brandboat) [reply test/e2e/framework/helpers/k8s/k8s_utils.go:393]: Agree with Chia-Ping's point. I also believe that we should invoke `DeleteWorkloadAndPods` in the `tearDownNamespace` method.

Comment (chia7712) [reply test/e2e/framework/helpers/k8s/k8s_utils.go:393]: @chenyulin0719 any feedback?

Comment (chenyulin0719) [reply test/e2e/framework/helpers/k8s/k8s_utils.go:393]: @chia7712 
I'm working on extending TearDownNamespace() to cover below namespace level object:
1. Deployment
2. DaemonSet
3. GetReplicaSet
4. GetStatefulSet
5. Pods
6. PVC

>PV, PVC, pods, and jobs should be cleanup when the e2e is closing even if the e2e does not use such resource. 

My previous thought is resource cleanup should be 'explicit'.
ex: "If I create pvc in e2e test, I should delete it after this e2e test.  If I don't use pvc in e2e test,  I don't need to check pvc existing.

However, currently, we're not using the 'explicit' cleanup in every e2e test, and looks like TearDownNamespace() is the shared function to cleanup namespace for multiple e2e test. So I'll follow the suggestion and make it more generic.

@brandboat Thanks for the suggestion. I've checked the DeleteWorkloadAndPods() only deletes one workload object per trigger. (If we have 10 Jobs in namespace, we would need to call DeleteWorkloadAndPods() 10 times). Therefore, I will create a new function for each resource type.

Comment (chia7712) [reply test/e2e/framework/helpers/k8s/k8s_utils.go:393]: @chenyulin0719 thanks for update. sounds good!

Comment (chenyulin0719) [reply test/e2e/framework/helpers/k8s/k8s_utils.go:453]: New commit pushed, 
Also clenaup Deployment,DaemonSet,ReplicaSet,Stateful,PVC in TearDownNamespace()

File: test/e2e/framework/helpers/k8s/k8s_utils.go [hunk 2]
```diff
@@ -391,7 +451,38 @@ func (k *KubeCtl) DeleteNamespace(namespace string) error {
...
-	err := k.DeletePods(namespace)
+	var err error
+	err = k.DeleteDeployments(namespace)
```
Comment (chia7712) [inline test/e2e/framework/helpers/k8s/k8s_utils.go:455]: @chenyulin0719 Could you add comments to explain the order of delete?

**Review by craigcondit — APPROVED** [review]

File: test/e2e/framework/helpers/k8s/k8s_utils.go (hunk 2 above)
Comment (craigcondit) [reply test/e2e/framework/helpers/k8s/k8s_utils.go:455]: I think this is fairly self-explanatory. Objects that create other objects are deleted first.

**Pull request closed without merging**